src/safepoints.rs

```diff
@@ -0,0 +1,138 @@
+extern crate ykstackmaps;
+
+use std::collections::HashMap;
+use std::path::Path;
+use ykstackmaps::{LocKind, LocOffset, SMRec, StackMapParser};
+
+use core::mem;
+
+#[derive(Clone, Copy, Hash, PartialEq, Eq, Debug)]
+pub struct ReturnAddress(pub u64);
+
+/// The offset from the stack pointer.
+#[derive(Debug)]
+struct SPO(u32);
+
+/// A `PtrSlot` identifies a stack root at a given safepoint using its offset
+/// from the Stack Pointer.
+///
+/// A base pointer (not to be confused with X86 terminology, where base pointer
+/// refers to the frame pointer register) is a pointer an object. In opposition
+/// to this, a derived pointer points to the interior of an object.
+///
+/// The Derived variant of a `PtrSlot` also contains a Stack Pointer offset to
+/// the base of the object.
+#[derive(Debug)]
+enum PtrSlot {
+    Base(SPO),
+    Derived(SPO, SPO),
+}
+
+/// Contains root locations for a Safepoint
+#[derive(Debug)]
+pub struct SafepointRoots {
+    /// A list of registers which contain roots across a safepoint
+    // DWARF Register number mapping can be found here:
+    // Pg.63 https://software.intel.com/sites/default/files/article/402129/mpx-linux64-abi.pdf
+    registers: Vec<u16>,
+
+    /// A list of `PtrSlot`s which correspond to roots accessible from a stack
+    /// pointer offset across a safepoint.
+    stack_offsets: Vec<PtrSlot>,
+}
+
+/// Pointer slots in stackmap locations are treated as offsets from a
+/// particular register. In the case of a Direct or Indirect location kind,
+/// these are the frame pointer and stack pointer registers respectively. To
+/// avoid calculating this during a GC pause, we convert all offsets to be
+/// from an SP upfront.
+fn as_sp_offset(offset: &LocOffset) -> SPO {
+    match offset {
+        LocOffset::I32(ref o) => SPO(*o as u32),
+        _ => panic!("Offset must be signed"),
+    }
+}
+
+fn gen_safepoint_roots(stackmap: SMRec) -> SafepointRoots {
+    // The first 2 locations are uninteresting, however, they should be constants.
+    let mut idx = 2;
+    debug_assert_eq!(
+        mem::discriminant(&stackmap.locs[0].kind),
+        mem::discriminant(&LocKind::Constant)
+    );
+    debug_assert_eq!(
+        mem::discriminant(&stackmap.locs[1].kind),
+        mem::discriminant(&LocKind::Constant)
+    );
+
+    // The 3rd location specifies the number of de-opt locations. De-opt
+    // params are not interesting to us, so we skip over them
+    let num_deopts = {
+        match stackmap.locs[idx].offset {
+            LocOffset::U32(c) => c,
+            _ => panic!("Constants must be u32"),
+        }
+    };
+    idx += (num_deopts as usize) + 1;
```

Comment:
Laurie, would you recommend `try_from` here?

Jake's `as` is safe, but IIRC, a `try_from` up-cast from `u32` to `usize` would be free on x86_64.

Reply:
Sadly, yes, `u32` to `usize` isn't guaranteed to be safe :/ There are then two options:

1) `try_from`
2) Somewhere have an assert along the lines of `assert!(usize::max_value() > u32::max_value())`.

Reply:
> u32 to usize isn't guaranteed to be safe

It is for x86_64. I was under the assumption this isn't portable code, as we are already talking about x86 registers.

Reply:
The LLVM statepoint API is only compatible with x86_64, so the cast is safe in this case. Whether we should be future-proofing it for other architectures is another question. As far as I'm aware, there are no plans yet to port the statepoint API to other architectures, although the changes would be mostly mechanical.

Reply:
Why don't we use a try_from and be done with it? Presumably we won't be repeatedly loading the stackmaps section in, so it should be a small one-time cost.

Reply:
I'm fine with assuming 64 bits if a) we say so somewhere b) have an assert to enforce that. The disaster zone is that one day stackmaps are ported to (say) a 16 bit (or 128 bit) architecture and this crate silently starts doing something weird.

Reply:
That's true in this case, but there will be other places in the future where these sorts of casts will be necessary and the cost would then be incurred on each collection. In addition, a lot of the stack walking code is platform specific and not portable.

Reply:
@ltratt A disclaimer seems like a good idea .

Reply:
I really think we should have an `assert` or similar in that checks this somewhere that's guaranteed to be called at least once so that if someone does run it on an inappropriate machine it goes splat. 

A possibly good approach is to use https://crates.io/crates/static_assertions -- `assert_eq_size` might be the right thing for us.

Reply:
Ah, sorry I missed this one. I'll add an assertion at the beginning of the table construction function.

Reply:
OK.